algorithms/linfa-pls/src/pls_svd.rs

```diff
@@ -0,0 +1,181 @@
+use crate::errors::{PlsError, Result};
+use crate::utils;
+use linfa::{dataset::Records, traits::Fit, traits::Transformer, DatasetBase, Float};
+use ndarray::{s, Array1, Array2, ArrayBase, Data, Ix2};
+use ndarray_linalg::{svd::*, Lapack, Scalar};
+
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+#[derive(Debug, Clone)]
+pub struct PlsSvdParams {
+    n_components: usize,
+    scale: bool,
+}
+
+impl PlsSvdParams {
+    pub fn new(n_components: usize) -> PlsSvdParams {
+        PlsSvdParams {
+            n_components,
+            scale: true,
+        }
+    }
+
+    pub fn scale(mut self, scale: bool) -> Self {
+        self.scale = scale;
+        self
+    }
+}
+
+impl Default for PlsSvdParams {
+    fn default() -> Self {
+        Self::new(2)
+    }
+}
+
+impl<F: Float + Scalar + Lapack, D: Data<Elem = F>> Fit<'_, ArrayBase<D, Ix2>, ArrayBase<D, Ix2>>
+    for PlsSvdParams
+{
+    type Object = Result<PlsSvd<F>>;
+
+    fn fit(
+        &self,
+        dataset: &DatasetBase<ArrayBase<D, Ix2>, ArrayBase<D, Ix2>>,
+    ) -> Result<PlsSvd<F>> {
+        if dataset.nsamples() < 2 {
+            return Err(PlsError::NotEnoughSamplesError(format!(
+                "should be greater than 1, got {}",
+                dataset.records().nsamples()
+            )));
+        }
+        // we'll compute the SVD of the cross-covariance matrix = X.T.dot(Y)
+        // This matrix rank is at most min(n_samples, n_features, n_targets) so
+        // n_components cannot be bigger than that.
+
+        let rank_upper_bound = dataset
+            .nsamples()
+            .min(dataset.nfeatures())
+            .min(dataset.targets().ncols());
+        if 1 > self.n_components || self.n_components > rank_upper_bound {
+            return Err(PlsError::BadComponentNumberError(format!(
+                "n_components should be in [1, {}], got {}",
+                rank_upper_bound, self.n_components
+            )));
+        }
+        let (x, y, x_mean, y_mean, x_std, y_std) =
+            utils::center_scale_dataset(&dataset, self.scale);
+
+        // Compute SVD of cross-covariance matrix
+        let c = x.t().dot(&y);
```

Comment:
you can also use here [ndarray_linalg::lobpcg::lobpcg](https://github.com/rust-ndarray/ndarray-linalg/blob/master/ndarray-linalg/src/lobpcg/lobpcg.rs#L515) which is matrix free and has therefore a lower complexity

Reply:
I suggest to open an issue with this one and the usage of `TruncatedSVD` to be addressed in a future PR about benchmarking and improving PLS.